src/ephemeris.c

```diff
@@ -68,6 +70,129 @@ static s8 calc_sat_state_xyz(const ephemeris_t *e, const gps_time_t *t,
+ *  ICD 5.1: A.3.1.2, with corrections from RTCM 3.2 p.186
+ *
+ * \param ydot Pointer to output array
+ * \param pos Pointer to position input array
+ * \param vel Pointer to velocity input array
+ * \param acc Pointer to acceleration input array
+ * \param e Pointer to GLO ephemeris
+ */
+static void calc_ydot(double ydot[6],
+                      const double pos[3],
+                      const double vel[3],
+                      const double acc[3])
+{
+
+  double r = sqrt(pow(pos[0], 2) +
+                  pow(pos[1], 2) +
+                  pow(pos[2], 2));
+
+  ydot[0] = vel[0];
+  ydot[1] = vel[1];
+  ydot[2] = vel[2];
+
+  ydot[3] = -GPS_GM / pow(r,3) * pos[0]
+            - 3/2 * GLO_J02 * GPS_GM * pow(GLO_A_E, 2) / pow(r, 5) *
+              pos[0] * (1 - 5 * pow(pos[2], 2) / pow(r, 2))
+            + pow(GPS_OMEGAE_DOT, 2) * pos[0]
+            + 2 * GPS_OMEGAE_DOT * vel[1]
+            + acc[0];
+
+  ydot[4] = -GPS_GM / pow(r,3) * pos[1]
+            - 3/2 * GLO_J02 * GPS_GM * pow(GLO_A_E, 2) / pow(r, 5) *
+              pos[1] * (1 - 5 * pow(pos[2], 2) / pow(r, 2))
+            + pow(GPS_OMEGAE_DOT, 2) * pos[1]
+            + 2 * GPS_OMEGAE_DOT * vel[0]
+            + acc[1];
+
+  ydot[5] = -GPS_GM / pow(r,3) * pos[2]
+            - 3/2 * GLO_J02 * GPS_GM * pow(GLO_A_E, 2) / pow(r, 5) *
+              pos[2] * (3 - 5 * pow(pos[2], 2) / pow(r, 2))
+            + acc[2];
+}
+
+/** Calculate satellite position, velocity and clock offset from GLO ephemeris.
+ *
+ * \param e Pointer to an ephemeris structure for the satellite of interest
+ * \param t time at which to calculate the satellite state
+ * \param pos Array into which to write calculated satellite position [m]
+ * \param vel Array into which to write calculated satellite velocity [m/s]
+ * \param clock_err Pointer to where to store the calculated satellite clock
+ *                  error [s]
+ * \param clock_rate_err Pointer to where to store the calculated satellite
+ *                       clock error [s/s]
+ *
+ * \return  0 on success,
+ *         -1 if ephemeris is not valid or too old
+ */
+static s8 calc_sat_state_glo(const ephemeris_t *e, const gps_time_t *t,
+                             double pos[3], double vel[3],
+                             double *clock_err, double *clock_rate_err)
+{
+  assert(e != NULL);
+  assert(t != NULL);
+  assert(pos != NULL);
+  assert(vel != NULL);
+  assert(clock_err != NULL);
+  assert(clock_rate_err != NULL);
+
+  double dt = abs(gpsdifftime(t, &e->toe));
+
+  if (dt > 900) {
+    log_error("GLO: Integration end point is not within 900 s of TOE");
+    return 1;
+  }
+
+  if (ceil(dt / GLO_NUM_STEPS)) {
+    double ydot0[6], yn[6];
+    double h = dt / GLO_NUM_STEPS;
+
+    calc_ydot(ydot0, e->glo.pos, e->glo.vel, e->glo.acc);
+    memcpy(&yn[0], e->glo.pos, sizeof(double) * 3);
+    memcpy(&yn[3], e->glo.vel, sizeof(double) * 3);
+
+    /* Runge-Kutta integration algorithm */
+    for (u8 i = 0; i < GLO_NUM_STEPS; i++) {
```

Comment:
Instead of constant number of steps, we should have constant maximum step size (see earlier comment), and the number of steps and actual step size (which will always be less) computed here as:

num_steps = ceil(dt / GLO_MAX_STEP_LENGTH);
h = dt / num_steps;